dev/io.openliberty.org.apache.myfaces.4.0/src/org/apache/myfaces/util/WebXmlParser.java

```diff
@@ -0,0 +1,312 @@
+            {
+                // Quarkus
+                webXml = toDocument(ClassUtils.getCurrentLoader(WebXmlParser.class).getResource("META-INF/web.xml"));
+            }
+            
+            if (webXml != null)
+            {
+                return parseErrorPages(webXml.getDocumentElement());
+            }
+        }
+        catch (Throwable e)
+        {
+            LOGGER.log(Level.SEVERE, "Could not load or parse web.xml", e);
+        }
+
+        return null;
+    }
+
+    private static Map<String, String> getWebFragmentXmlsErrorPages(ExternalContext context)
+    {
+        Map<String, String> webFragmentXmlsErrorPages = null;
+
+        try
+        {
+            Enumeration<URL> webFragments = ClassUtils.getContextClassLoader()
+                    .getResources("META-INF/web-fragment.xml");
+            while (webFragments.hasMoreElements())
+            {
+                try
+                {
+                    URL url = webFragments.nextElement();
+                    Document webFragmentXml = toDocument(url);
+                    if (webFragmentXml != null)
+                    {
+                        if (webFragmentXmlsErrorPages == null)
+                        {
+                            webFragmentXmlsErrorPages = parseErrorPages(webFragmentXml.getDocumentElement());
+                        }
+                        else
+                        {
+                            Map<String, String> temp = parseErrorPages(webFragmentXml.getDocumentElement());
+                            for (Map.Entry<String, String> entry : temp.entrySet())
+                            {
+                                if (!webFragmentXmlsErrorPages.containsKey(entry.getKey()))
+                                {
+                                    webFragmentXmlsErrorPages.put(entry.getKey(), entry.getValue());
+                                }
+                            }
+                        }
+                    }
+                }
+                catch (Throwable e)
+                {
+                    LOGGER.log(Level.SEVERE, "Could not load or parse web-fragment.xml", e);
+                }
+            }
+        }
+        catch (IOException e)
+        {
+            LOGGER.log(Level.SEVERE, "Could not get web-fragment.xml from ClassLoader", e);
+        }
+
+        return webFragmentXmlsErrorPages;
+    }
+
+    private static Document toDocument(URL url) throws Exception
+    {
+        InputStream is = null;
+
+        try
+        {
+            // web.xml is optional
+            if (url == null)
+            {
+                return null;
+            }
+
+            if (System.getSecurityManager() != null)
+            {
+                is = (InputStream) AccessController.doPrivileged(
+                        (PrivilegedExceptionAction<Object>) () -> url.openStream());
+            }
+            else
+            {
+                is = url.openStream();
+            }
```

Comment:
Is there a JIRA for this yet? I can't recall.

Reply:
No, not yet wanted to get a review before opening a JIRA. I'll do that next and update the commit messages so it is clear what JIRA we're overlaying here. I won't merge until it is accepted in MyFaces

Reply:
https://issues.apache.org/jira/browse/MYFACES-4540